docs/proposals/high-availability.md

```diff
@@ -0,0 +1,34 @@
+For a very simple example of proposed behavior see: 
+* https://github.com/rrati/etcd-ha
+* go get github.com/rrati/etcd-ha
+
+In HA, the apiserver will be a gateway to etcd. It will provide an api for becoming master, updating the master lease, and releasing the lease.  This api is daemon agnostic, so to become the master the client will need to provide the daemon type and the lease duration when attemping to become master.  The apiserver will attempt to create a key in etcd based on the daemon type that contains the client's hostname/ip and port information. This key will be created with a ttl from the lease duration provided in the request.  Failure to create this key means there is already a master of that daemon type, and the error from etcd will propigate to the client.  Successfully creating the key means the client making the request is the master.  When updating the lease, the apiserver will update the existing key with a new ttl.  The location in etcd for the HA keys is TBD.
+
+Leader election is first come, first serve.  The first daemon of a specific type to request leadership will become the master.  All other daemons of that type will fail until the current leader releases the lease or fails to update the lease within the expiration time.  On startup, all daemons should attempt to become master.  The daemon that succeeds is the master and should perform all functions of that daemon.  The daemons that fail to become the master should not perform any tasks and sleep for their lease duration and then attempt to become the master again.
+
+The daemon that becomes master should create a Go routine to manage the lease.  This process should be created with a channel that the main daemon process can use to release the master lease.  Otherwise, this process will update the lease and sleep, waiting for the next update time or notification to release the lease.  If there is a failure to update the lease, this process should force the entire daemon to exit.  Daemon exit is meant to prevent potential split-brain conditions.  Daemon restart is implied in this scenario, by either the init system (systemd), or possible watchdog processes.  (See Design Discussion Notes)
+
+## Options added to daemons with HA functionality
+Some command line options would be added to daemons that can do HA:
+
+* Lease Duration - How long a daemon can be master
+
+* Number of Missed Lease Updates - How many updates can be missed before the lease as the master is lost
+
+## Design Discussion Notes on Scheduler/Controller
+Some daemons, such as the controller-manager, may fork numerous go routines to perform tasks in parallel.  Trying to keep track of all these processes and shut them down cleanly is untenable.  If a master daemon loses leadership then the whole daemon should exit with an exit code indicating that the daemon is not the master.  The daemon should be restarted by a monitoring system, such as systemd, or a software watchdog.
```

Comment:
I know we're written in go, but goroutine vs thread is an implementation detail, and in theory controllers could be written in any language. Therefore suggest you just say "thread". Also the "monitoring system" is going to be kubelet in many (most?) configurations.

> Trying to keep track of all these processes and shut them down cleanly is untenable.

I don't think this is a great reason. I would phrase this paragraph like this:

```
Some daemons may run numerous threads in order to perform tasks in parallel. Upon losing master status, such daemons should exit instantly instead of attempting to gracefully shut down such threads. This is to ensure that, in the case there's some propagation delay in informing the threads they should stop, the lame-duck threads won't interfere with the new master.

Since all daemons will be run by kubelet or some other monitoring system, this will just result in a restart.
```